plugins/analysis-icu/src/main/java/org/elasticsearch/index/mapper/ICUCollationKeywordFieldMapper.java

```diff
@@ -0,0 +1,738 @@
+
+        @Override
+        public Query nullValueQuery() {
+            if (nullValue() == null) {
+                return null;
+            }
+            return termQuery(nullValue(), null);
+        }
+
+        @Override
+        public IndexFieldData.Builder fielddataBuilder() {
+            failIfNoDocValues();
+            return new DocValuesIndexFieldData.Builder();
+        }
+
+        @Override
+        protected BytesRef indexedValueForSearch(Object value) {
+            if (value == null) {
+                return null;
+            }
+            if (value instanceof BytesRef) {
+                value = ((BytesRef) value).utf8ToString();
+            }
+
+            if (collator != null) {
+                RawCollationKey key;
+                synchronized (collatorLock) {
+                    key = collator.getRawCollationKey(value.toString(), null);
+                }
+
+                return new BytesRef(key.bytes, 0, key.size);
+            } else {
+                throw new IllegalStateException("collator is null");
+            }
+        }
+
+        public static DocValueFormat COLLATE_FORMAT = new DocValueFormat() {
+            @Override
+            public String getWriteableName() {
+                return "collate";
+            }
+
+            @Override
+            public void writeTo(StreamOutput out) throws IOException {
+            }
+
+            @Override
+            public String format(long value) {
+                throw new UnsupportedOperationException();
+            }
+
+            @Override
+            public String format(double value) {
+                throw new UnsupportedOperationException();
+            }
+
+            @Override
+            public String format(BytesRef value) {
+                int encodedLength = IndexableBinaryStringTools.getEncodedLength(value.bytes, value.offset, value.length);
+                char[] encoded = new char[encodedLength];
+                IndexableBinaryStringTools.encode(value.bytes, value.offset, value.length, encoded, 0, encodedLength);
+                return new String(encoded, 0, encodedLength);
+            }
+
+            @Override
+            public long parseLong(String value, boolean roundUp, LongSupplier now) {
+                throw new UnsupportedOperationException();
+            }
+
+            @Override
+            public double parseDouble(String value, boolean roundUp, LongSupplier now) {
+                throw new UnsupportedOperationException();
+            }
+
+            @Override
+            public BytesRef parseBytesRef(String value) {
+                char[] encoded = value.toCharArray();
+                int decodedLength = IndexableBinaryStringTools.getDecodedLength(encoded, 0, encoded.length);
+                byte[] decoded = new byte[decodedLength];
+                IndexableBinaryStringTools.decode(encoded, 0, encoded.length, decoded, 0, decodedLength);
+                return new BytesRef(decoded);
```

Comment:
Maybe format and parseBytesRef should reuse the collator? Otherwise the keys you would get eg. in aggregations would make little sense. This requires additional serialization logic to serialize the configuration of the collator, you can look at `DocValueFormat.DateTime` to see how to make it work.